homeassistant/components/lcn/__init__.py

```diff
@@ -44,62 +35,89 @@
-            await hass.async_create_task(connection.async_connect(timeout=15))
-            connections.append(connection)
-            _LOGGER.info('LCN connected to "%s"', connection_name)
+            await lcn_connection.async_connect(timeout=15)
+        except pypck.connection.PchkAuthenticationError:
+            _LOGGER.warning('Authentication on PCHK "%s" failed.', config_entry.title)
+            return False
+        except pypck.connection.PchkLicenseError:
+            _LOGGER.warning(
+                'Maximum number of connections on PCHK "%s" was '
+                "reached. An additional license key is required.",
+                config_entry.title,
+            )
+            return False
         except TimeoutError:
-            _LOGGER.error('Connection to PCHK server "%s" failed', connection_name)
+            _LOGGER.warning('Connection to PCHK "%s" failed.', config_entry.title)
             return False
 
-    hass.data[DATA_LCN][CONF_CONNECTIONS] = connections
-
-    # load platforms
-    for component, conf_key in (
-        ("binary_sensor", CONF_BINARY_SENSORS),
-        ("climate", CONF_CLIMATES),
-        ("cover", CONF_COVERS),
-        ("light", CONF_LIGHTS),
-        ("scene", CONF_SCENES),
-        ("sensor", CONF_SENSORS),
-        ("switch", CONF_SWITCHES),
-    ):
-        if conf_key in config[DOMAIN]:
+        _LOGGER.info('LCN connected to "%s"', config_entry.title)
+        hass.data[DOMAIN][config_entry.entry_id] = {
+            CONNECTION: lcn_connection,
+        }
+        config_entry.add_update_listener(async_entry_updated)
+
+        # remove orphans from entity registry which are in ConfigEntry but were removed
+        # from configuration.yaml
+        if config_entry.source == config_entries.SOURCE_IMPORT:
+            entity_registry = await er.async_get_registry(hass)
+            entity_registry.async_clear_config_entry(config_entry.entry_id)
+
+        # forward config_entry to components
+        for component in PLATFORMS:
             hass.async_create_task(
-                async_load_platform(
-                    hass, component, DOMAIN, config[DOMAIN][conf_key], config
-                )
+                hass.config_entries.async_forward_entry_setup(config_entry, component)
             )
 
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Close connection to PCHK host represented by config_entry."""
+    # forward unloading to platforms
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok and config_entry.entry_id in hass.data[DOMAIN]:
+        host = hass.data[DOMAIN].pop(config_entry.entry_id)
+        await host[CONNECTION].async_close()
+
+    return unload_ok
+
+
+async def async_setup(hass, config):
```

Comment:
Please keep `async_setup` above `async_setup_entry` in the module. That's standard.